### PR TITLE
feat(proxy): master domain routing for remote apps/services

### LINKED ISSUE #8668: [Bug]: Domain routing breaks when app is deployed on remote server without public ingress.
### Error Message and Logs

I am personally working on this issue.

When Coolify is running the main proxy (Traefik) on a public VPS and a second server is connected via WireGuard (10.8.0.X/24), deploying an app to the remote server and assigning a domain does not work if DNS points only to the public VPS.

Expected behavior
Request to https://<domain> hits the public VPS proxy and is forwarded over WireGuard to the service running on the remote server.

Actual behavior
The domain resolves to the public VPS, but the proxy cannot reach/route to the service on the remote server, so the site returns 404/502/timeout.

Topology
Server A (public): Coolify + Traefik proxy, public IP: x.x.x.x 
Server B (remote): connected via WireGuard, private IP: 10.8.0.X
DNS: all domains point to Server A

### Steps to Reproduce

1. Add Server B as a remote server reachable only via WireGuard (10.8.0.x).
2. Deploy any HTTP app to Server B (e.g. nginx / hello-world) and assign any domain (or wildcard domain)
3. Ensure DNS A record for <domain> points to Server A.
4. Visit https://<domain>.

### Example Repository URL

_No response_

### Coolify Version

v4.0.0-beta.463

### Are you using Coolify Cloud?

No (self-hosted)

### Operating System and Version (self-hosted)

Ubuntu 22.04

### Additional Information

I’m working on this. 

Plan: implement edge proxy routes to remote tunnel IP by generating Traefik file-provider config on the proxy host using the remote server’s WG IP + published host port. I’ll open a draft PR once I tests that pass.

### PR DESCRIPTION
**Pull request by Twest2**
## Changes

- TLDR: Added the option to have a master domain router like a Kubernetes cluster. With this setup, `*.example.com` points to one public VPS, which acts as the single entry point for incoming traffic. Private servers at home or elsewhere connect to that VPS over ssh (including over a private network like WireGuard) and the VPS forwards each request to the correct private server. This lets private servers stay off the public internet while still being reachable through normal domain names. The traffic goes through the main server and gets forwarded via a load balancer to the private server. The private server only has to have an ssh connection to the public server.

- Added an optional team-level Master Domain Router setting so a single public server can act as the entry point for remote/private servers.
- This makes it possible to point `*.example.com` at one public VPS, forward traffic over a private network such as WireGuard, and keep the destination servers off the public internet.
- Added `EdgeProxyRemoteRouteService` to generate edge Traefik routes for remote applications and services.
- Edge now writes stable dynamic config files at:
- `/data/coolify/proxy/dynamic/service-remote-<service-uuid>.yaml`
- `/data/coolify/proxy/dynamic/application-remote-<application-uuid>.yaml`
- Generated config includes:
- HTTP to HTTPS redirect
- HTTPS router with `certResolver=letsencrypt`
- load balancer target `http://<remote_host>:<published_host_port>`
- Route files are updated on deploy and redeploy, and removed when the resource is deleted.
- Missing or invalid host/port values are skipped with warnings instead of breaking the whole routing update.
- Added master-routing support for remote database proxies.
- Added test coverage for:
- `EdgeProxyRemoteRouteServiceTest`
- `DatabaseProxyMasterRoutingTest`
- `ServerMasterDomainRouterSettingTest`

## Issues

Closes #8668
- Remote servers connected via Wireguard would not get traffic just by adding it as a server.

## Category

- [X] Bug fix
- [x] New feature

## Public VPS
image

## Private server
image

## AI Assistance

- [X] AI was used (please describe below)

**If AI was used:**

- Tools used: Codex
- How extensively: I used ChatGPT's Codex 5.4 to help identify the issue and what needed to be done in the future. I also had it write all test cases and resolve the merge conflicts that came up.

## Testing

1. Ran the new unit test file:
   - `php artisan test tests/Unit/EdgeProxyRemotePortForwardServiceTest.php tests/Unit/EdgeProxyRemoteRouteServiceTest.php tests/Unit/DatabaseProxyMasterRoutingTest.php`
   - Result: `60 passed (269 assertions)`

Verified by tests:

- Ran: `php artisan test tests/Unit/DeleteServiceTest.php tests/Unit/DeleteResourceJobTest.php tests/Unit/EdgeProxyRemoteRouteServiceTest.php tests/Unit/EdgeProxyRemotePortForwardServiceTest.php tests/Feature/EdgeProxyServerResolverTest.php tests/Feature/ServerMasterDomainRouterSettingTest.php`
- Result:
- `60 passed (269 assertions)`
- Verified by tests:
- Traefik dynamic config is generated with HTTP to HTTPS redirect support.
- HTTPS routers use `certResolver=letsencrypt`.
- Services route to `http://<remote_host>:<published_host_port>`.
- Route files are created and updated during deploy/redeploy and removed on resource deletion.
- Route file names stay stable per application or service UUID.
- Remote application TCP and UDP routing continues to work.
- Invalid routing inputs are skipped with actionable warnings instead of failing hard.

Tested on my own personal setup with one VPS 3 remote servers. Tested Minecraft servers, Nextcloud deployments, and VERT.

## Contributor Agreement

> [!IMPORTANT]
>
> - [X] I have read and understood the [contributor guidelines](https://github.com/coollabsio/coolify/blob/v4.x/CONTRIBUTING.md). If I have failed to follow any guideline, I understand that this PR may be closed without review.
> - [X] I have searched [existing issues](https://github.com/coollabsio/coolify/issues) and [pull requests](https://github.com/coollabsio/coolify/pulls) (including closed ones) to ensure this isn't a duplicate.
> - [X] I have tested all the changes thoroughly with a local development instance of Coolify and I am confident that they will work as expected when a maintainer tests them.

### CODE REVIEW

**Review by Cinzya — CHANGES_REQUESTED** [review]
Not necessarily a bad idea, but it comes with a few flaws / oversights.

From what I can tell, you are right now only covering service routes, but not applications. So as soon as you deploy a application, it won't work. The same probably applies for database proxies as well, would need to check the code again to confirm.

There are users who use Coolify actually the opposite way. Have the Coolify dashboard on a non-public, private network, while every other remote server is public.
Also one server as the entrypoint makes it a single point of failure.
It would probably be better if this is a setting people can opt-in from the UI, not make it the default behavior for every user.

I probably missed a few more points, I didn't actually test the implementation myself, just glanced over the code. Also make sure to actually do a integration test, aka test with a running Coolify installation, not just running the Unit test files.

**Comment by Twest2** [comment]
I see your point about the applications, I'll make sure that that is resolved and I didn't think about people having coolify the opposite way. I'll look into this and also add a button in the server config that will let users opt into this.

File: app/Actions/Database/StartDatabaseProxy.php [hunk 1]
```diff
@@ -24,15 +25,15 @@ public function handle(StandaloneRedis|StandalonePostgresql|StandaloneMongodb|St
     {
         $databaseType = $database->database_type;
         $network = data_get($database, 'destination.network');
-        $server = data_get($database, 'destination.server');
+        $deploymentServer = data_get($database, 'destination.server');
```
Comment (Twest2) [inline app/Actions/Database/StartDatabaseProxy.php:32]: Two different roles used to be used on this variable. I changed it so that its "where the DB container actually runs". I found it easy to misuse and master routing needs both servers tracked at once.

File: app/Actions/Database/StartDatabaseProxy.php [hunk 2]
```diff
@@ -66,61 +92,68 @@ public function handle(StandaloneRedis|StandalonePostgresql|StandaloneMongodb|St
     stream {
        server {
             listen $database->public_port;
-            proxy_pass $containerName:$internalPort;
+            proxy_pass $upstreamTarget;
```
Comment (Twest2) [inline app/Actions/Database/StartDatabaseProxy.php:101]: See new line 72

File: app/Actions/Database/StopDatabaseProxy.php [hunk 3]
```diff
@@ -22,16 +23,69 @@ class StopDatabaseProxy
 
     public function handle(StandaloneRedis|StandalonePostgresql|StandaloneMongodb|StandaloneMysql|StandaloneMariadb|StandaloneKeydb|ServiceDatabase|StandaloneDragonfly|StandaloneClickhouse $database)
     {
-        $server = data_get($database, 'destination.server');
+        $deploymentServer = data_get($database, 'destination.server');
```
Comment (Twest2) [inline app/Actions/Database/StopDatabaseProxy.php:30]: Again just a name change to make more sense with the new changes.

File: app/Jobs/ApplicationDeploymentJob.php [hunk 4]
```diff
@@ -508,6 +509,22 @@ private function post_deployment()
...
+                    $this->application_deployment_queue->addLogEntry("Edge proxy routing warning: {$warning}", 'stderr');
+                }
+            } catch (\Throwable $e) {
+                \Log::warning('Failed to sync edge proxy route for application deployment '.$this->deployment_uuid.': '.$e->getMessage());
+                $this->application_deployment_queue->addLogEntry(
+                    'Edge proxy routing warning: Failed to sync edge proxy route configuration. Check master domain server settings and edge proxy connectivity.',
+                    'stderr'
+                );
+            }
+        }
```
Comment (Twest2) [inline app/Jobs/ApplicationDeploymentJob.php:559]: this makes app deploys also update edge/master-domain routing, without breaking the deploy if routing sync fails.

File: app/Livewire/Server/Show.php [hunk 5]
```diff
@@ -45,9 +46,17 @@ class Show extends Component
 
     public bool $isBuildServer;
 
+    public bool $isMasterDomainRouterEnabled;
```
Comment (Twest2) [inline app/Livewire/Server/Show.php:50]: The master server is where all the domains are pointed to. The master server is then responsible for routing them to where it needs to go.

File: app/Livewire/Server/Show.php [hunk 6]
```diff
@@ -492,6 +526,34 @@ public function submit()
...
+            ->where('id', '!=', $this->server->id)
+            ->whereRelation('settings', 'is_master_domain_router_enabled', true)
+            ->exists();
+
+        if (! $masterDomainServerExists) {
+            return;
+        }
+
+        $this->isMasterDomainRouterLocked = true;
+        $this->masterDomainRouterLockMessage = 'Disabled because another server in this team is already selected as the master domain router. Disable it there first.';
```
Comment (Twest2) [inline app/Livewire/Server/Show.php:571]: Don't let multiple servers be the "master" server

File: resources/views/livewire/server/show.blade.php [hunk 7]
```diff
@@ -286,6 +286,20 @@ class="w-full input opacity-50 cursor-not-allowed"
...
+                                        label="Enable master domain routing?"
+                                        helper="{{ $masterDomainRouterLockMessage }}" />
+                                @else
+                                    <x-forms.checkbox canGate="update" :canResource="$server"
+                                        id="isMasterDomainRouterEnabled" label="Enable master domain routing?"
+                                        helper="Enable this on exactly one server when all domains point to this server and it should forward traffic to resources running on other servers (for example over VPN / WireGuard)."
+                                        :disabled="$isValidating" />
+                                @endif
+                            </div>
+                        @endif
```
Comment (Twest2) [inline resources/views/livewire/server/show.blade.php:303]: For the settings, the little information thingy that tells users what a master server is.

File: tests/Unit/EdgeProxyRemoteRouteServiceTest.php [hunk 8]
```diff
@@ -0,0 +1,1081 @@
...
+
+    $deploymentServer = Mockery::mock(Server::class)->makePartial();
+    $deploymentServer->id = 37;
+    $deploymentServer->ip = '10.8.0.37';
+    $deploymentServer->proxy = ['type' => 'NONE'];
+
+    $application = new Application;
+    $application->uuid = 'application-unsupported-protocol';
+    $application->build_pack = 'nixpacks';
+    $application->fqdn = 'https://valid-app.example.com:3000,tcp://minecraft.example.com:25565';
```
Comment (Twest2) [inline tests/Unit/EdgeProxyRemoteRouteServiceTest.php:1473]: minecraft :)

**Comment by Twest2** [comment]
Howdy @Cinzya, I added support for applications and databases. I also added in a "master server" option in the server config that enables or disables this feature. I did implementation test on my machine and it works. I also added some more test cases via the command below. I had Codex write some of these test cases just because I couldn't get every edge case.

`php artisan test tests/Feature/ServerMasterDomainRouterSettingTest.php tests/Unit/DatabaseProxyMasterRoutingTest.php tests/Unit/EdgeProxyRemoteRouteServiceTest.php`

**Comment by Twest2** [comment]
I found an error when I implemented this on my on VPS running coolify. TLS certificates would not route properly whenever the port on the secondary server wasn't resolved. I have implemented this PR on my own VPS and have seen that it has worked. I have tested it with the one click deploy of VERT which does work (with https). I just tried it on the one minecraft server and unfortunately it didn't work which I'm working on currently. Edit: Works.

**Comment by Twest2** [comment]
Howdy @Cinzya, I have fixed the issue I found and tested both http/https and tcp applications running on a live VPS and a secondary home server.

I tested the "Master router" feature where only one server has all the DNS records set to it and only one server can enable this feature. From there, I deployed the one click service VERT which ran as expected and I also ran the one click minecraft server aswell and it worked as expected.  
I have not tested out a database simply because I don't know how to test out a database in a production environment but I am confident that it will work. 

How I had my system set up is the main vps with *.twestbrook.com pointed to it and my home server connected via wireguard over ssh. From there, the end user types in the domain that they want and it gets routed to the appropriate server after passing through the master server, or in my case, to the vps then to the home server.

Please let me know if you have any questions!

**Review by Twest2 — COMMENTED** [review]
Tested with web apps (VERT) and applications (minecraft) on a live vps. Works as expected.

File: config/constants.php
Comment (Twest2) [inline config/constants.php:1]: Line 5 

`'version' => '4.0.0-beta.465',` must be changed.

**Comment by Twest2** [comment]
@Cinzya  and @ShadowArcanist, would either of you mind reviewing this PR when you get a chance? I know it is a large PR so I completely understand that it may take some time.

I currently have it running on my personal VPS with other servers connected to it and it has been working in my testing/use case. The overall idea is somewhat similar to a Kubernetes cluster.

I did not want to ping Peak or Andras since I know they are busy with the recent updates, but I would really appreciate it if one of yall had time to look it over.

**Comment by ShadowArcanist** [comment]
> @Cinzya and @ShadowArcanist, would either of you mind reviewing this PR when you get a chance? I know it is a large PR so I completely understand that it may take some time.
> 
> I currently have it running on my personal VPS with other servers connected to it and it has been working in my testing/use case. The overall idea is somewhat similar to a Kubernetes cluster.
> 
> I did not want to ping Peak or Andras since I know they are busy with the recent updates, but I would really appreciate it if one of yall had time to look it over.

Sorry, I’m unable to review this PR because the changes are beyond my current level of familiarity with the codebase.

**Comment by Twest2** [comment]
I readded one of the features in the new update which was a merge conflict. I readded the "Server Details" feature in servers.

**Comment by Twest2** [comment]
@coderabbitai do your thing!

**Comment by Twest2** [comment]
@coderabbitai please check the most recent commit and double check that i hit all your points.

**Comment by Twest2** [comment]
I confirmed this still works after the recent Traefik update.

**Comment by Twest2** [comment]
@andrasbacsai and/or @peaklabs-dev and/or @Cinzya, would either of you mind reviewing this PR when you get a chance? I know it is a large PR so I completely understand that it may take some time.

I currently have it running on my personal VPS with other servers connected to it and it has been working in my testing/use case. The overall idea is somewhat similar to a Kubernetes cluster.

**Comment by ShadowArcanist** [comment]
Hey @Iisyourdad thanks for the PR!

Unfortunately I have to close this after taking a closer look at the code changes.

Here is why:

1. You are making all server traffic go through one single server which gives less control for the user and creates a single point of failure for every application.


2. The routing is auto-derived by parsing each app's docker-compose ports: and env var defaults at runtime. That's fragile — port changes, unusual compose layouts, and variable expansion can silently break routing, and users only find out from log warnings after the fact.

3. The edge router forwards directly to the container's published host port over the tunnel and bypasses the remote server's Traefik entirely. The remote server's proxy can no longer manage its own certs, middlewares, or routing rules, and a specific TLS split (master terminates, remote stays plaintext) gets forced on every application.

4. Container labels are silently rewritten based on team state — shouldUsePublicCertResolver() strips certresolver=letsencrypt from apps on every non-master server. Flipping one checkbox retroactively changes the label output of every app on every server in the team, which is a lot of action-at-a-distance for one setting.

5. It's a team-wide on/off switch with no per-app control. Users can't mix "route this app through the gateway, deploy this other one directly."

The direction I'd like to take instead is user-driven gateway routes: a UI where you add domain, target URL, path prefix, middlewares (HTTPS redirect / strip prefix ) and Coolify writes the matching Traefik dynamic config file behind the scenes. 

The remote server's Traefik stays in charge of its own containers, routes are scoped per-route (not team-wide), container labels are untouched, and routes can point at anything — Coolify-managed server or not. I've opened #9600 for that.

None of this is a comment on the effort you put in. I just don't want to ship the auto-everything team-wide version when the dynamic-config-with-UI approach solves the same problem without the footguns.

Thanks again for putting this together 🙏

**Comment by Twest2** [comment]
Hey @ShadowArcanist those are fair points, I'll look at the other PR and see how to make a better version.

**Comment by Twest2** [comment]
Hey @ShadowArcanist Im wondering if we could revisited this and try to combine for both your implementation and mine aswell. I have done some edits to it since this pr has been closed which allows for the user to opt a service out of master domain routing (which isn't shown in the pr but in the image below.)

I feel like this is a feature that a lot of people would use, but I also see the point of having your implementation aswell. Could we combine both features in and have the end user have both? I can fairly easily transfer your implementation over and also add support for udp and tcp packets aswell which my feature has.

The reason why I bring this up is just due to my server architecture (and other peoples server architecture that I have seen online) which forces me to have this feature inorder to use Coolify. I have this running on my own personal VPS and it works great.

I would add the gateway section in that you're doing and have both features. 


Screenshot 2026-04-29 154145
